sycl/doc/extensions/proposed/sycl_ext_oneapi_device_image_content.asciidoc

```diff
@@ -0,0 +1,254 @@
+
+Available on when the compiler is {cpp}20 or higher.
+
+_Returns:_ A view of the raw bytes for this device image.
+The data behind this view has the same lifetime as the `kernel_bundle` that
+contains this `device_image`.
+
+'''
+
+
+== Device image format for {dpcpp}
+
+This section is non-normative and applies only to the {dpcpp} implementation.
+The format of the data returned by `device_image::ext_oneapi_get_content` and
+`device_image::ext_oneapi_get_content_view` depends on the backend of the kernel
+bundle that contains the device image and also on the `State` of the device
+image.
+
+=== Format on Level Zero
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The format could change in the future, but it is
+  currently a SPIR-V module representing one or more kernels.
+  This SPIR-V module may be partially linked, with references to internal
+  library functions that are not defined in the SPIR-V module.
+
+* `bundle_state::object`: The same format as `bundle_state::input`.
+
+* `bundle_state::executable`: The device image content is native ISA for the
+  device, which can be passed to `zeModuleCreate` as `ZE_MODULE_FORMAT_NATIVE`
+  format.
+
+:ref1: ../proposed/sycl_ext_oneapi_free_function_kernels.asciidoc#level-zero-and-opencl-compatibility
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+Level Zero module that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on OpenCL
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The format could change in the future, but it is
+  currently a SPIR-V module as described above for Level Zero.
+
+* `bundle_state::object`: The device image content is an unspecified format that
+  is created by calling `clCompileProgram`.
+
+* `bundle_state::executable`: The device image content is executable binary
+  device code representing one or more kernels, which can be passed to
+  `clCreateProgramWithBinary`.
+
+[_Note:_ The interface to kernels in the device image is not defined in the
+general case, which means there is no portable way to invoke kernels from a
+OpenCL `cl_program` object that is created from the raw device image content.
+However, see link:{ref1}[here] for a limited case where this portability is
+guaranteed.
+_{endnote}_]
+
+=== Format on CUDA
+
+The format depends on the `State` of the device image:
+
+* `bundle_state::input`: The device image content is a PTX module representing
+  one or more kernels.
+
+* `bundle_state::object`: ???
+
+* `bundle_state::executable`: The device image content is a CUBIN module
+  representing one or more kernels.
+
+
+== Example
+
+:ref2: ../proposed/sycl_ext_oneapi_free_function_kernels.asciidoc
+
+A kernel bundle can contain multiple device images with different
```

Comment:
Is there any way to get information about device for a given device image?

Reply:
Not directly.  You can indirectly ask if a device image is compatible with a device via `device_image::has_kernel(kernel_id, device)`.  There are a couple reasons why it's hard to provide a direct query:

* For device images that are not in `executable` state, the device image could be compatible with many different devices.  For example, a SPIR-V module can be compiled for many different devices, including future devices that don't even exist yet.

* Even for `executable` device images, the image could be compatible with many different devices.  For example, I think this is the case for Nvidia CUBIN modules.

I think we could add a query like `device_image::is_compatible(device)` if that would help.  This would be similar to `device_image::has_kernel(kernel_id, device)`, except it asks about the device image a whole, rather than a specific kernel.

Reply:
if there is device_image::has_kernel(kernel_id, device), then I think it is fine.